typing/typecore.ml

```diff
@@ -4776,18 +4857,23 @@ and type_function ?in_function loc attrs env (expected_mode : expected_mode)
                                           ty_fun,
                                           explanation)))
   in
-  if has_local then
-    eqmode ~loc ~env arg_mode Alloc_mode.local
-      (Param_mode_mismatch ty_expected');
+  if has_local || has_unique then begin
+    let mode = match has_local, has_unique with
+      | true, true -> Mode.Alloc.local_unique
+      | true, false -> Mode.Alloc.local
```

Comment:
I think that this case (and the one below) is too restrictive. It should probably have a variable for its uniqueness. That will allow code like the following (maybe put this in the testsuite):

```ocaml
let foo : local_ unique_ string -> unit = fun (local_ s) -> ()
let bar : local_ unique_ string -> unit = fun (unique_ s) -> ()
```

The additional variable (and the `assert false` below) could also be avoided by pulling the equality constraint up into these branches and using more specific equality functions (e.g. using equality on just locality rather than the whole mode).